client/extensions/woocommerce/state/ui/review-replies/selectors.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * External dependencies
+ */
+import { get, isObject, merge } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { getReviewReply } from 'woocommerce/state/sites/review-replies/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Number|Object} The ID of the review reply (or object placeholder, if a new reply)
+ */
+export const getCurrentlyEditingReviewReplyId = ( state, siteId = getSelectedSiteId( state ) ) => {
+	return get( state, [ 'extensions', 'woocommerce', 'ui', 'reviewReplies', siteId, 'edits', 'currentlyEditingId' ], null );
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Number|Object} The ID of the review that a reply edit is associated with.
+ */
+export const getCurrentlyEditingReviewId = ( state, siteId = getSelectedSiteId( state ) ) => {
+	return get( state, [ 'extensions', 'woocommerce', 'ui', 'reviewReplies', siteId, 'edits', 'reviewId' ], null );
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Object} The local edits made to the reply.
+ */
+export const getReviewReplyEdits = ( state, siteId = getSelectedSiteId( state ) ) => {
+	return get( state, [ 'extensions', 'woocommerce', 'ui', 'reviewReplies', siteId, 'edits', 'changes' ], {} );
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Object} The reply merged with changes, or just the changes if a new reply
+ */
+export const getReviewReplyWithEdits = ( state, siteId = getSelectedSiteId( state ) ) => {
+	const reviewId = getCurrentlyEditingReviewId( state, siteId );
+	const replyId = getCurrentlyEditingReviewReplyId( state, siteId );
+	const reviewEdits = getReviewReplyEdits( state, siteId );
+
+	if ( isObject( replyId ) ) {
+		return { ...reviewEdits, id: replyId, parent: reviewId };
```

Comment:
When I first saw the object being placed under the `id` attribute here that made me pause a bit, but can see in the tests how it is used to track the placeholder. Is this a common pattern we are using? In the past I have just dropped something like `temporary-ID` in the `id` in the state tree and update once the object has been persisted by the API.

Reply:
Yeah this is a pattern we are currently using for at least orders, products, and variations. For new items (that don't have a real ID yet) we'll generate a unique placeholder ID (`thing_#`) and store that in the state tree. Once the object is saved to the UI we do get an ID back.